### PR TITLE
Refactor CRM_Contact_BAO_Relationship::add to autoload missing params from existing record

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
This is a pre-requisite to https://github.com/civicrm/civicrm-core/pull/15030.  The BAO add method needed updating so it loads missing parameters directly using standard methods rather than requiring that they are all passed in to the function.

Before
----------------------------------------
Have to pass a lot of parameters to the `CRM_Relationship_BAO_Relationship::add()` function.  Existing parameters not found by the function.

After
----------------------------------------
Per other standard functions, `add()` uses standard methods to find an existing record and use the params directly.  This only happens if an ID is passed into the function.

Technical Details
----------------------------------------
Use standard DAO find() method.

Comments
----------------------------------------

### CODE REVIEW

File: CRM/Contact/BAO/Relationship.php [hunk 1]
```diff
@@ -295,33 +295,38 @@ public static function add($params, $ids = [], $contactId = NULL) {
     if ($params['id']) {
       $hook = 'edit';
     }
-    //@todo hook are called from create and add - remove one
```
Comment (demeritcowboy) [inline CRM/Contact/BAO/Relationship.php]: This comment is still true if coming from legacyCreateMultiple which is still used in some places. I'm not suggesting to deal with that now, just that there is still a todo.

**Comment by demeritcowboy** [comment]
Thanks @mattwire - have taken a quick look and looks good so far. Give me a bit and I'll do a bit more.

**Comment by eileenmcnaughton** [comment]
@mattwire I'd rather only do the load if we are missing  specific params we need for the next processing steps - this could be part of a big batch update.

File: CRM/Contact/BAO/Relationship.php (hunk 1 above)
Comment (mattwire) [reply CRM/Contact/BAO/Relationship.php]: ok - I've moved the comment to where it belongs in `legacyCreateMultiple()`

**Comment by mattwire** [comment]
> @mattwire I'd rather only do the load if we are missing specific params we need for the next processing steps - this could be part of a big batch update.

@eileenmcnaughton Ok I've done that

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
This seems ok to me - will see if @demeritcowboy has further comments

**Comment by demeritcowboy** [comment]
* General standards
    * [`r-explain`] __PASS__
    * [`r-user`] __PASS__
    * [`r-doc`] __PASS__
    * [`r-run`] __PASS__
      * Looks good.
      * It's pre-existing but the head of household thing doesn't work because 6 is the wrong id, at least on the test site here. I'm not sure that feature would work properly anyway as written in all cases.
      * I don't currently see a way to test in the UI the situation where you haven't passed in enough params, which is partly what this was about. See also r-maint below.
* Developer standards
    * [`r-tech`] __PASS__
    * [`r-code`] __PASS__
    * [`r-maint`] __PASS__
      * I notice there are no existing tests on the add() function directly. The api relationship-create is well-tested, so indirectly add() is tested, just as we know the api does more manipulation after calling add(), so some of those tests might be really testing that later code not add(). But it seems tested enough that it shouldn't hold this up. I don't mind writing a test about the lack of params to demonstrate the before and after, but you don't have to wait for me.
    * [`r-test`] __PASS__

**Comment by demeritcowboy** [comment]
See test at https://github.com/civicrm/civicrm-core/pull/15121

**Comment by mattwire** [comment]
> It's pre-existing but the head of household thing doesn't work because 6 is the wrong id, at least on the test site here. I'm not sure that feature would work properly anyway as written in all cases.

I didn't like that there was a hardcoded ID there - but resisted the urge to fix because I couldn't find a  an existing example and.. more PRs.. but as it's *actually* broken on your site here's a PR: #15123

**Comment by demeritcowboy** [comment]
Thanks @mattwire. While I've always been a huge fan of the number 6 it's time for it to go. It's the wrong id on dmaster too and any stock new install.

**Pull request merged**